### DISCUSSION TITLE
TypedDict unions narrowing with literals - expected behaviour?

### DISCUSSION THREAD

**Discussion by mwaddoups** [microsoft/pyright#5435] [Q&A]
I'm seeing odd behaviour with how Literal narrowing is applied. See below

```python
from typing import TypedDict, Literal

class A(TypedDict): 
    name: Literal["A"]
    foo: str

class BC(TypedDict):
    name: Literal["B", "C"]
    bar: str

AorBC = A | BC

def get_name() -> Literal["C", "D"]:
    # ... some conditional to return C or D
    return "C"

def some_func(lst: list[AorBC]):
    name = get_name()   # type is correct as Literal["C", "D"]
    return [
        x for x in lst
        if x['name'] == name
        and x['bar'] == 'something'   # raises "bar" is not a defined key in A
    ]
```

This works fine if e.g. I change `get_name` to return `Literal["C"]`.

I would expect the type to get narrowed to `BC` here - not sure if this is a bug?

**Comment by erictraut** [comment]
This isn't a bug per se. This type narrowing pattern currently works only with a single literal value on the RHS of the `==` operator. It doesn't work with unions of literals. Incidentally, mypy also has this limitation in its implementation. 

What you're doing here is quite unusual. Typically, this pattern is used with a literal string expression such as `x['name'] == "C"`.

It's possible to add support for unions of literals as a feature request, but I'd want to understand your use case a bit more before committing to doing this. We may also want to wait to see if this is something that other pyright users want or need. If it's unique to your code, it's unlikely that we'd implement it.

**Reply by mwaddoups** [reply]
That makes sense and thanks for getting back so quickly on this!

My use case is essentially as described - I have an existing codebase (financial) where we work with a list of financial instruments that have four different `kind` values, and the `kind` values are linked to what keys you can expect. 

I suspect that in my specific case a much easier solution would be to write some type guards to go along with the types, and use those in cases where I need to narrow.

Our current codebase has a lot of `if kind in ('B', 'C', 'D')` which isn't a supported narrowing pattern in pyright anyway, so if I'm having to rewrite sections than writing a function like

```python
Instrument = Union[ .. ] # these different kinds of instruments
def is_bcd(inst: Instrument) -> TypeGuard[InstrumentB | InstrumentCD]
    inst['kind'] in ('B', 'C', 'D')
```

will allow to support this specific case (of course, without the benefit of external checks).

What would be cool is if we could specify custom acceptable type guard narrowing patterns for types we use so we don't have to touch existing code - but obviously well beyond pyright's scope.

**Reply by erictraut** [reply]
Thanks for the additional information about your use case. I've added support for unions of literals in the type narrowing code path, so your code sample above now works. This will be included in the next release.
